src/main/java/hudson/tasks/SMTPAuthentication.java

```diff
@@ -1,41 +1,142 @@
+import hudson.model.Item;
+import hudson.security.ACL;
+import hudson.util.ListBoxModel;
 import hudson.util.Secret;
+import jenkins.model.Jenkins;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+import org.kohsuke.stapler.AncestorInPath;
 import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.QueryParameter;
+
+import java.util.Collections;
+import java.util.UUID;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.StreamSupport;
 import hudson.Util;
 
 /**
  * @author <a href="mailto:nicolas.deloof@gmail.com">Nicolas De Loof</a>
  */
 public class SMTPAuthentication extends AbstractDescribableImpl<SMTPAuthentication> {
 
-    private String username;
+    private static final Logger LOGGER = Logger.getLogger(SMTPAuthentication.class.getName());
+
+    /** use StandardUsernamePasswordCredentials instead */
+    @Deprecated
+    private transient String username;
+
+    /** use StandardUsernamePasswordCredentials instead */
+    @Deprecated
+    private transient Secret password;
 
-    private Secret password;
+    /** The ID of the Jenkins Username/Password credential to use. */
+    private String credentialsId;
 
     @DataBoundConstructor
-    public SMTPAuthentication(String username, Secret password) {
+    public SMTPAuthentication(String credentialsId) {
+        this.credentialsId = credentialsId;
+    }
+
+    @Deprecated
+    @Restricted(NoExternalUse.class)
+    SMTPAuthentication(String username, Secret password) {
         this.username = Util.fixEmptyAndTrim(username);
         this.password = password;
+        readResolve();
     }
 
+    @Deprecated
     public String getUsername() {
         return username;
     }
 
+    @Deprecated
     public Secret getPassword() {
         return password;
     }
 
+    public String getCredentialsId() {
+        return credentialsId;
+    }
+
+    private Object readResolve() {
+        if (username != null || password != null) {
+            LOGGER.log(Level.CONFIG, "Migrating the Mailer SMTP authentication details to credential...");
+
+            final CredentialsStore store = StreamSupport.stream(CredentialsProvider.lookupStores(Jenkins.get()).spliterator(), false)
+                    .filter(s -> s instanceof SystemCredentialsProvider.StoreImpl)
+                    .findFirst()
+                    .orElseThrow(() -> new RuntimeException("Could not migrate the Mailer SMTP authentication details to credential, as the system credentials provider was missing."));
+
+            final boolean isSuccess = Retryable.retry(9, (attempt) -> {
+                LOGGER.log(Level.CONFIG, "Attempt {0}...", attempt);
+
+                final String id = UUID.randomUUID().toString();
+
+                final StandardUsernamePasswordCredentials migratedCredential = new UsernamePasswordCredentialsImpl(
+                        CredentialsScope.GLOBAL,
+                        id,
```

Comment:
you can pass null as the id, then when adding to the store the provider will generate one for you.
After adding you can just get the id with
this.credentialsId = migratedCredential.getId();

Reply:
Will the `migratedCredential` object created with the null `id` have its id property updated in-place when the SystemCredentialsProvider saves it? Or would it have to be retrieved again from the provider after the save?

Reply:
The Constructor already takes care for generating an id if null is passed. This will then never change.
You might want to look up the code in the email-ext plugin that did the same migration not very long ago. They even look for existing credentials with same user/password and reuse it (https://github.com/jenkinsci/email-ext-plugin/blob/6ee9208dcec866749edbdc12f135159ae7cf0c17/src/main/java/hudson/plugins/emailext/MailAccount.java#L260)

Reply:
Implemented the suggestion